### PR TITLE
Add z_listaccounts API.

### LINKED ISSUE #5635: Add z_listaccounts RPC
Currently no RPC exists for listing all the accounts created through `z_getnewaccount`.

### PR DESCRIPTION
**Pull request by therealyingtong**
Closes #5635

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3423,6 +3423,70 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+        auto ufvkMetadata = ufvkMetadataPair->second;
+        auto diversifiersMap = ufvkMetadata.GetKnownReceiverSetsByDiversifierIndex();
+
+        auto ufvkInfo = pwalletMain->GetUnifiedFullViewingKey(ufvkId);
+
+        for (auto it = diversifiersMap.begin(); it != diversifiersMap.end(); it++) {
+            UniValue address(UniValue::VOBJ);
+
+            auto j = it->first;
+            address.pushKV("diversifier", j.ToString());
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3476]:
```suggestion
            UniValue jVal;
            jVal.setNumStr(ArbitraryIntStr(std::vector(j.begin(), j.end())));
            addrEntry.pushKV("diversifier_index", jVal);
```

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3423,6 +3423,70 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    KeyIO keyIO(Params());
+    UniValue ret(UniValue::VARR);
+
+    auto hdChain = pwalletMain->GetMnemonicHDChain();
+    auto num_accounts = hdChain->GetAccountCounter();
+
+    for (int acctId = 0; acctId < num_accounts; acctId++) {
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3457]: Why iterate over the account counter rather than just iterate over `mapUnifiedAccountKeys` directly?

**Review by nuttycom — CHANGES_REQUESTED** [review]

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3423,6 +3423,71 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+
+    LOCK(pwalletMain->cs_wallet);
+
+    KeyIO keyIO(Params());
+    UniValue ret(UniValue::VARR);
+
+    auto hdChain = pwalletMain->GetMnemonicHDChain();
+    auto accountsMap = pwalletMain->mapUnifiedAccountKeys;
+
+    for (auto accountMap = accountsMap.begin(); accountMap != accountsMap.end(); accountMap++) {
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3457]:
```suggestion
    for (const auto& [acctKey, ufvkId] : pwalletMain->mapUnifiedAccountKeys) {
```

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -3423,6 +3423,71 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+        UniValue account(UniValue::VOBJ);
+
+        uint64_t acctId = accountMap->first.second;
+        account.pushKV("account", acctId);
+
+        // Get the receivers for this account.
+        auto ufvkId = accountMap->second;
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3464]:
```suggestion
```

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -3423,6 +3423,71 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+        auto ufvkMetadataPair = pwalletMain->mapUfvkAddressMetadata.find(ufvkId);
+        auto ufvkMetadata = ufvkMetadataPair->second;
+        auto diversifiersMap = ufvkMetadata.GetKnownReceiverSetsByDiversifierIndex();
+
+        auto ufvkInfo = pwalletMain->GetUnifiedFullViewingKey(ufvkId);
+
+        for (auto it = diversifiersMap.begin(); it != diversifiersMap.end(); it++) {
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3471]:
```suggestion
        for (const auto& [j, receiverTypes] : diversifiersMap) {
```

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -3423,6 +3423,71 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+            UniValue addrEntry(UniValue::VOBJ);
+
+            auto j = it->first;
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3474]:
```suggestion
```

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -3423,6 +3423,71 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+            UniValue jVal;
+            jVal.setNumStr(ArbitraryIntStr(std::vector(j.begin(), j.end())));
+            addrEntry.pushKV("diversifier_index", jVal);
+
+            auto uaPair = std::get<std::pair<UnifiedAddress, diversifier_index_t>>(ufvkInfo->Address(it->first, it->second));
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3479]:
```suggestion
            auto uaPair = std::get<std::pair<UnifiedAddress, diversifier_index_t>>(ufvk.Address(j, receiverTypes));
```

File: src/wallet/rpcwallet.cpp [hunk 8]
```diff
@@ -3423,6 +3423,71 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+        uint64_t acctId = accountMap->first.second;
+        account.pushKV("account", acctId);
+
+        // Get the receivers for this account.
+        auto ufvkId = accountMap->second;
+        auto ufvkMetadataPair = pwalletMain->mapUfvkAddressMetadata.find(ufvkId);
+        auto ufvkMetadata = ufvkMetadataPair->second;
+        auto diversifiersMap = ufvkMetadata.GetKnownReceiverSetsByDiversifierIndex();
+
+        auto ufvkInfo = pwalletMain->GetUnifiedFullViewingKey(ufvkId);
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3469]:
```suggestion
        auto ufvk = pwalletMain->GetUnifiedFullViewingKey(ufvkId).value();
```

**Review by nuttycom — CHANGES_REQUESTED** [review]
It looks like the test needs to be completed; also a few small style suggestions.

File: qa/rpc-tests/wallet_z_listaccounts.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    nuparams,
+    start_nodes,
+    wait_and_assert_operationid_status,
+)
+
+class ZListAccountsTest (BitcoinTestFramework):
+    def __init__(self):
+        super().__init__()
+        self.num_nodes = 3
+        self.setup_clean_chain = True
```
[the remark this answers is not in the document]
Comment (therealyingtong) [reply qa/rpc-tests/wallet_z_listaccounts.py:15]: I didn't mean to add this test actually -- I'll just inline it in some other test.

**Review by nuttycom — APPROVED** [review]
utACK

File: qa/rpc-tests/wallet_accounts.py [hunk 10]
```diff
@@ -58,16 +63,21 @@ def run_test(self):
         # Verify that just creating the account does not generate any visible addresses
         addresses = self.nodes[0].z_listaddresses()
         assert_equal([], addresses)
+        accounts = self.nodes[0].z_listaccounts()
+        assert_equal(accounts[0]['account'], 0)
```
Comment (daira) [inline qa/rpc-tests/wallet_accounts.py:67]: This and the checks below should probably check the length of `accounts`, to make sure there are no spurious ones.

File: src/wallet/rpcwallet.cpp [hunk 11]
```diff
@@ -3447,6 +3447,69 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
...
+
+    if (fHelp || params.size() > 1)
+        throw runtime_error(
+            "z_listaccounts\n"
+            "\nReturns the list of accounts created through z_getnewaccount.\n"
+            "\nResult:\n"
+            "[\n"
+            "   {\n"
+            "     \"account\": \"uint\",           (uint) The account id\n"
+            "        {\n"
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:3463]: This isn't valid JSON syntax, I think it should be something like `"addresses": [{...}, ...]`.

**Review by daira — CHANGES_REQUESTED** [review]
The error in the help text is blocking.

**Review by daira — APPROVED** [review]
utACK

**Pull request merged**